File: IAIso-v5.0/core/iaiso-php/src/Conformance/PolicyRunner.php
<?php

declare(strict_types=1);

namespace IAIso\Conformance;

use IAIso\Policy\Policy;
use IAIso\Policy\PolicyException;
use IAIso\Policy\PolicyLoader;

/** @internal */
final class PolicyRunner
{
    public const TOLERANCE = 1e-9;

    private function __construct() {}

    /** @return VectorResult[] */
    public static function run(string $specRoot): array
    {
        $file = $specRoot . '/policy/vectors.json';
        $doc = json_decode((string) file_get_contents($file), true, 512, JSON_THROW_ON_ERROR);
        $out = [];
        foreach ($doc['valid'] ?? [] as $v) {
            $out[] = self::runValid($v);
        }
        foreach ($doc['invalid'] ?? [] as $v) {
            $out[] = self::runInvalid($v);
        }
        return $out;
    }

    private static function runValid(array $v): VectorResult
    {
        $name = 'valid/' . $v['name'];
        try {
            $p = PolicyLoader::build($v['document']);

            if (isset($v['expected_pressure']) && is_array($v['expected_pressure'])) {
                $err = self::checkPressure($p, $v['expected_pressure']);
                if ($err !== null) return VectorResult::fail('policy', $name, $err);
            }
            if (isset($v['expected_consent']) && is_array($v['expected_consent'])) {
                $err = self::checkConsent($p, $v['expected_consent']);
                if ($err !== null) return VectorResult::fail('policy', $name, $err);
            }
            if (isset($v['expected_metadata']) && is_array($v['expected_metadata'])) {
                if (count($v['expected_metadata']) !== count($p->metadata)) {
                    return VectorResult::fail('policy', $name,
                        'metadata size: got ' . count($p->metadata)
                        . ', want ' . count($v['expected_metadata']));
                }
            }
            return VectorResult::pass('policy', $name);
        } catch (\Throwable $e) {
            return VectorResult::fail('policy', $name, 'build failed: ' . $e->getMessage());
        }
    }

    private static function checkPressure(Policy $p, array $ep): ?string
    {
        $checks = [
            ['token_coefficient', $p->pressure->tokenCoefficient],
            ['tool_coefficient', $p->pressure->toolCoefficient],
            ['depth_coefficient', $p->pressure->depthCoefficient],
            ['dissipation_per_step', $p->pressure->dissipationPerStep],
            ['dissipation_per_second', $p->pressure->dissipationPerSecond],
            ['escalation_threshold', $p->pressure->escalationThreshold],
            ['release_threshold', $p->pressure->releaseThreshold],
        ];
        foreach ($checks as [$k, $got]) {
            if (isset($ep[$k])) {
                $want = (float) $ep[$k];
                if (abs($got - $want) > self::TOLERANCE) {
                    return "$k: got $got, want $want";
                }
            }
        }
        if (isset($ep['post_release_lock'])) {
            $want = (bool) $ep['post_release_lock'];
            if ($want !== $p->pressure->postReleaseLock) {
                return 'post_release_lock mismatch';
            }
        }
        return null;
    }

    private static function checkConsent(Policy $p, array $ec): ?string
    {
        if (array_key_exists('issuer', $ec)) {
            $want = $ec['issuer'];
            if ($want !== $p->consent->issuer) {
                $g = $p->consent->issuer ?? 'null';
                $w = $want ?? 'null';
                return "consent.issuer: got $g, want $w";
            }
        }
        if (isset($ec['default_ttl_seconds'])) {
            $want = (float) $ec['default_ttl_seconds'];
            if (abs($p->consent->defaultTtlSeconds - $want) > self::TOLERANCE) {
                return "default_ttl_seconds: got {$p->consent->defaultTtlSeconds}, want $want";
            }
        }
        if (isset($ec['required_scopes'])) {
            if (count($ec['required_scopes']) !== count($p->consent->requiredScopes)) {
                return 'required_scopes length mismatch';
            }
        }
        if (isset($ec['allowed_algorithms'])) {
            if (count($ec['allowed_algorithms']) !== count($p->consent->allowedAlgorithms)) {
                return 'allowed_algorithms length mismatch';
            }
        }
        return null;
    }

    private static function runInvalid(array $v): VectorResult
    {
        $name = 'invalid/' . $v['name'];
        $expectPath = $v['expect_error_path'];
        try {
            PolicyLoader::build($v['document']);
            return VectorResult::fail('policy', $name,
                "expected error containing '$expectPath', got Ok");
        } catch (PolicyException $e) {
            if (!str_contains($e->getMessage(), $expectPath)) {
                return VectorResult::fail('policy', $name,
                    "expected error containing '$expectPath', got: " . $e->getMessage());
            }
            return VectorResult::pass('policy', $name);
        }
    }
}

File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Conformance/PolicyRunner.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json.Nodes;
using Iaiso.Policy;

namespace Iaiso.Conformance;

internal static class PolicyRunner
{
    public static List<VectorResult> Run(string specRoot)
    {
        var path = Path.Combine(specRoot, "policy", "vectors.json");
        var doc = JsonNode.Parse(File.ReadAllBytes(path))!.AsObject();
        var results = new List<VectorResult>();
        if (doc["valid"] is JsonArray valid)
            foreach (var v in valid) results.Add(RunValid(v!.AsObject()));
        if (doc["invalid"] is JsonArray invalid)
            foreach (var v in invalid) results.Add(RunInvalid(v!.AsObject()));
        return results;
    }

    private static VectorResult RunValid(JsonObject v)
    {
        string name = "valid/" + v["name"]!.GetValue<string>();
        try
        {
            var p = PolicyLoader.Build(v["document"]);
            if (v["expected_pressure"] is JsonObject ep)
            {
                string? err = CheckPressure(p, ep);
                if (err is not null) return VectorResult.Fail("policy", name, err);
            }
            if (v["expected_consent"] is JsonObject ec)
            {
                string? err = CheckConsent(p, ec);
                if (err is not null) return VectorResult.Fail("policy", name, err);
            }
            if (v["expected_metadata"] is JsonObject em)
            {
                if (em.Count != p.Metadata.Count)
                    return VectorResult.Fail("policy", name,
                        $"metadata size: got {p.Metadata.Count}, want {em.Count}");
            }
            return VectorResult.Pass("policy", name);
        }
        catch (Exception e)
        {
            return VectorResult.Fail("policy", name, "build failed: " + e.Message);
        }
    }

    private static string? CheckPressure(Iaiso.Policy.Policy p, JsonObject ep)
    {
        string[] labels = {
            "token_coefficient", "tool_coefficient", "depth_coefficient",
            "dissipation_per_step", "dissipation_per_second",
            "escalation_threshold", "release_threshold"
        };
        double[] gots = {
            p.Pressure.TokenCoefficient, p.Pressure.ToolCoefficient,
            p.Pressure.DepthCoefficient, p.Pressure.DissipationPerStep,
            p.Pressure.DissipationPerSecond, p.Pressure.EscalationThreshold,
            p.Pressure.ReleaseThreshold,
        };
        for (int i = 0; i < labels.Length; i++)
        {
            if (ep[labels[i]] is JsonValue v && v.TryGetValue<double>(out var d))
            {
                if (Math.Abs(gots[i] - d) > ConformanceRunner.Tolerance)
                    return $"{labels[i]}: got {gots[i]}, want {d}";
            }
        }
        if (ep["post_release_lock"] is JsonValue prl && prl.TryGetValue<bool>(out var b))
        {
            if (b != p.Pressure.PostReleaseLock) return "post_release_lock mismatch";
        }
        return null;
    }

    private static string? CheckConsent(Iaiso.Policy.Policy p, JsonObject ec)
    {
        if (ec["issuer"] is not null)
        {
            string? want = ec["issuer"] is JsonValue iv && iv.TryGetValue<string>(out var s)
                ? s : null;
            if (want is null ? p.Consent.Issuer is not null : want != p.Consent.Issuer)
                return $"consent.issuer: got {p.Consent.Issuer}, want {want}";
        }
        if (ec["default_ttl_seconds"] is JsonValue tv && tv.TryGetValue<double>(out var t))
        {
            if (Math.Abs(p.Consent.DefaultTtlSeconds - t) > ConformanceRunner.Tolerance)
                return $"default_ttl_seconds: got {p.Consent.DefaultTtlSeconds}, want {t}";
        }
        if (ec["required_scopes"] is JsonArray rs && rs.Count != p.Consent.RequiredScopes.Count)
            return "required_scopes length mismatch";
        if (ec["allowed_algorithms"] is JsonArray aa && aa.Count != p.Consent.AllowedAlgorithms.Count)
            return "allowed_algorithms length mismatch";
        return null;
    }

    private static VectorResult RunInvalid(JsonObject v)
    {
        string name = "invalid/" + v["name"]!.GetValue<string>();
        string expectPath = v["expect_error_path"]!.GetValue<string>();
        try
        {
            PolicyLoader.Build(v["document"]);
            return VectorResult.Fail("policy", name,
                $"expected error containing '{expectPath}', got Ok");
        }
        catch (PolicyException e)
        {
            return e.Message.Contains(expectPath)
                ? VectorResult.Pass("policy", name)
                : VectorResult.Fail("policy", name,
                    $"expected error containing '{expectPath}', got: {e.Message}");
        }
    }
}
